src/System.Collections/tests/Generic/Dictionary/DictionaryTests.cs

```diff
@@ -0,0 +1,164 @@
+        {
+            TestCopyConstructor(size, keyValueSelector, dictionarySelector, comparer);
+        }
+
+        public static IEnumerable<object[]> CopyConstructorInt32Data
+        {
+            get
+            {
+                return GetCopyConstructorData(i => i);
+            }
+        }
+
+        public static IEnumerable<object[]> CopyConstructorStringData
+        {
+            get
+            {
+                var comparers = new IEqualityComparer<string>[]
+                {
+                    null,
+                    EqualityComparer<string>.Default,
+                    StringComparer.Ordinal,
+                    StringComparer.OrdinalIgnoreCase
+                };
+
+                return GetCopyConstructorData(i => i.ToString(), comparers);
+            }
+        }
+
+        private static IEnumerable<object[]> GetCopyConstructorData<T>(Func<int, T> keyValueSelector, IEqualityComparer<T>[] comparers = null)
+        {
+            var dictionarySelectors = new Func<Dictionary<T, T>, IDictionary<T, T>>[]
+            {
+                d => d,
+                d => new DictionarySubclass<T, T>(d),
+                d => new ReadOnlyDictionary<T, T>(d)
+            };
+
+            var sizes = new int[] { 0, 1, 2, 3, 4, 5, 10, 50 };
+
+            comparers = comparers ?? new IEqualityComparer<T>[]
+            {
+                null,
+                EqualityComparer<T>.Default
+            };
+
+            foreach (Func<Dictionary<T, T>, IDictionary<T, T>> dictionarySelector in dictionarySelectors)
+            {
+                foreach (int size in sizes)
+                {
+                    foreach (IEqualityComparer<T> comparer in comparers)
+                    {
+                        yield return new object[] { size, keyValueSelector, dictionarySelector, comparer };
+                    }
+                }
+            }
+        }
+
+        private static void TestCopyConstructor<T>(int size, Func<int, T> keyValueSelector, Func<Dictionary<T, T>, IDictionary<T, T>> dictionarySelector)
+        {
+            IDictionary<T, T> expected = CreateDictionary(size, keyValueSelector);
+            IDictionary<T, T> input = dictionarySelector(CreateDictionary(size, keyValueSelector));
+
+            Assert.Equal(expected, new Dictionary<T, T>(input));
+        }
+
+        private static void TestCopyConstructor<T>(int size, Func<int, T> keyValueSelector, Func<Dictionary<T, T>, IDictionary<T, T>> dictionarySelector, IEqualityComparer<T> comparer)
+        {
+            IDictionary<T, T> expected = CreateDictionary(size, keyValueSelector, comparer);
+            IDictionary<T, T> input = dictionarySelector(CreateDictionary(size, keyValueSelector, comparer));
+
+            Assert.Equal(expected, new Dictionary<T, T>(input, comparer));
+        }
+
+        private static Dictionary<T, T> CreateDictionary<T>(int size, Func<int, T> keyValueSelector, IEqualityComparer<T> comparer = null)
```

Comment:
Should this return IDictionary instead of Dictionary, and should the Funcs take an IDictionary instead of Dictionary as input?